File: 509. Fibonacci Number.java
lass Solution {
    Map<Integer,Integer> mem = new HashMap();
    public int fib(int N) {
        if(N == 0) return 0;
        if(N == 1) return 1;
        if(mem.get(N) != null) {
            return mem.get(N);
        }
        mem.put(N, fib(N-1) + fib(N-2));
        return mem.get(N);
    }
}

File: 509. Fibonacci Number.kt
class Solution {
    fun fib(n: Int): Int {
           if (n == 1) {
            return 1
        }
        var dp = IntArray(50)

        dp[1] = 1
        dp[2] = 1

        for (i in 3 until n + 1) {
            dp[i] = dp[i - 1] +  dp[i - 2]
        }
        return dp[n]
    }
}
